mh/time_utils.py

```diff
@@ -21,8 +21,72 @@
 
 import datetime
 
+
 EPOCH = datetime.datetime(1970, 1, 1)
+TICKS_PER_CYCLE = 23040  # 6.4 hours per cycle, 3.2 hours per daytime/nighttime
 
 
 def datetime_to_int(dt):
     return int((dt - EPOCH).total_seconds())
+
+SECONDS_PER_DAY = 24*60*60
+JHEN_EVENT_OFFSET = 14  # Cycle of 14 (real) days
+FOG_START = 0  
+JHEN_START = 1  # 1 day of fog
+JHEN_END = 3  # Followed by 2 (real) days of sandstorm
+
+
+def current_tick():
+    """
+    This product of this function is based on the assumption that
+    any decreasing integer value is sufficient for tracking the
+    passage of time in Monster Hunter Tri.
+    
+    Debugging has shown that the internal time value (initialized
+    via a request to the server which responds using this function)
+    continuously decreases by one per second, and an internal modulus
+    is used to turn the ever-decreasing value into a point in the
+    day/night cycle (with a cycle length specified by the server).
+    
+    It is also suspected that, should this value ever reach zero, an
+    integer underflow happens, which disrupts the day/night cycle.
+    
+    As a result, to ensure that the underflow never happens, the current
+    implementation uses a reversed epoch time that counts down from
+    UINT_MAX (the game's internal maximum for this value) as the seconds
+    pass. This also has the effect of time "passing" even when the
+    server is shut down.
+    """
+    return 0xFFFFFFFF - datetime_to_int(datetime.datetime.now())
+
+
+def current_server_time():
+    return datetime_to_int(datetime.datetime.now())
+
+
+def get_jhen_cycle_time(curr_time=None):
+    if curr_time is None:
+        curr_time = current_server_time()
+    current_server_day = int(curr_time//SECONDS_PER_DAY)
+    if (current_server_day % JHEN_EVENT_OFFSET) >= JHEN_END:
+        next_day = current_server_day + (JHEN_EVENT_OFFSET - (current_server_day % JHEN_EVENT_OFFSET))
+    else:
+        next_day = current_server_day - (current_server_day % JHEN_EVENT_OFFSET)
+    
+    return next_day*SECONDS_PER_DAY
```

Comment:
By having some intermediate variables, you'll make your intent clearer:
```python
def get_jhen_event_times():
    current_day = int(current_server_time() // SECONDS_PER_DAY)
    day_offset = current_day % JHEN_EVENT_OFFSET
    if day_offset < JHEN_END:
        period_start = (current_day - day_offset) * SECONDS_PER_DAY
    else:  # Upcoming period
        period_start = (current_day - day_offset + JHEN_EVENT_OFFSET) * SECONDS_PER_DAY
    return (
        period_start + FOG_START*SECONDS_PER_DAY,
        period_start + JHEN_START*SECONDS_PER_DAY,
        period_start + JHEN_END*SECONDS_PER_DAY
    )
```

Reply:
Done